posts/fall2019/js-chats-3/index.md

```diff
@@ -0,0 +1,1285 @@
+- [Final Challenge](#final-challenge)
+
+## Quick Review of the `class` Lifecycle Methods
+
+Before we dive into hooks, let's review some lifecycle methods
+in the old `class` syntax. 
+Here are some important ones that we need to know: 
+
+- `setState`: update the state within a component, triggers `render` 
+- `componentDidMount`: called when a component is just mounted.
+  `componentDidMount` is allowed to perform side-effect here such as API
+  calls, and state updates with `setState`
+- `render`: React calls `render` to get the updated version of the component
+- `componentDidUpdate`: called after the component re-rendered. Side effects allowed
+
+The lifecycle methods are well summarized in the following flow chart:
+
+![react lifecycle](images/lifecycle.png)
+
+> See the diagram at: http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/
+
+Exercise: Build a simple counter component, but the initial value 
+of the counter is given by the server. Here we emulate an API call
+to fetch the initial value.
+
+```jsx
+// faking a call to the server to get an initial value
+function getInitialCounterValue() {
+  return new Promise(resolve => setTimeout(() => resolve(3)), 200);
+}
+```
+
+<details>
+
+<summary> Click to see solution </summary>
+
+  ```jsx
+  class Counter extends React.Component {
+    state = { count: 0 }
+
+    constructor() {
+      this.onIncrement = this.onIncrement.bind(this);
+    }
+
+    componentDidMount() {
+      getInitialCounterValue().then(count => {
+        this.setState({ count });
+      });
+    }
+
+    onIncrement() {
+      this.setState(({ count }) => { count: count + 1 });
+    }
+
+    render() {
+      return (
+        <div>
+          <h3>{this.state.count}</h3>
+          <button onClick={this.onIncrement}>Increment</button>
+        </div>
+      );
+    }
+  }
+  ```
+
+</details>
+
+## Quick Review of Function Component
+
+For some simple component that has no state (__stateless component__), 
+using the class syntax can be quite "wordy".
+
+```jsx
+class MyDisplayInfoComponent extends React.Component {
+  render() {
+    return (
+      <ul>
+        <li>Name: {this.props.name}</li>
+        <li>Major: {this.props.major}</li>
+      </ul>
+    );
+  }
+}
+```
+
+To get reduce the amount of code we have to write,
+we can use functions to write component instead of class.
+
+```jsx
+function MyDisplayInfoComponent (props) {
+  return (
+    <ul>
+      <li>Name: {props.name}</li>
+      <li>Major: {props.major}</li>
+    </ul>
+  );
+}
+```
+
+To simplify it further, we can even use the ES6 arrow
+function syntax and parameter destructuring assignment.
+
+```jsx
+const MyDisplayInfoComponent = ({ name, major }) => (
+  <ul>
+    <li>Name: {props.name}</li>
+    <li>Major: {props.major}</li>
+  </ul>
+);
+```
+
+These stateless components are said to be _pure functions_.
+Pure functions are simply functions in programming that 
+obeys the property of a mathematical function:
```

Comment:
```suggestion
obey the property of a mathematical function:
```

Also known as idempotence in computer science (though in math idempotence is something else), btw.